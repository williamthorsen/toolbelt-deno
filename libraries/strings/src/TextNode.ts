import { segmentByDelimited } from './segmentByDelimited.ts';
import { splitDelimited } from './splitDelimited.ts';
import { IntSeededRng, pickInteger, Seed } from '../../numbers/mod.ts';
import { validateDelimiters } from './validateDelimiters.ts';

const DELIMIT = {
  opening: '[',
  closing: ']',
  separator: '|',
};

/**
 * Class for building an abstract syntax tree (AST) of static text nodes (tokens) and variants.
 * @experimental
 */
export abstract class TextNode {
  content: string;
  variantIndexCount = 0;

  protected constructor(content: string) {
    this.content = content;
  }

  static create(content: string): TextNode {
    if (TokenNode.hasDelimited(content)) {
      validateDelimiters(content, { ...DELIMIT, throwOnError: true });
    }
    if (TokenNode.isDelimited(content)) {
      return new VariantNode(content);
    } else {
      return new TokenNode(content);
    }
  }

  static decodeIndices(encodedIndices: string | VariantIndices): Integer[] {
    if (typeof encodedIndices !== 'string') {
      if (isNumberArray(encodedIndices)) return encodedIndices;
      return flatten(encodedIndices);
    }

    const regex = /\d+/g;
    let match;
    const indices = [];

    while ((match = regex.exec(encodedIndices)) !== null) {
      indices.push(Number(match[0]));
    }

    return indices;
  }

  static delimit(content: string): string {
    return enclose(content, DELIMIT);
  }

  static encodeIndices(arr: VariantIndices, depth = 0): string {
    let result = '';

    for (let i = 0; i < arr.length; i++) {
      const element = arr[i];
      if (typeof element === 'number') {
        result += element.toString();
      } else if (Array.isArray(element)) {
        let encodedIndices = TextNode.encodeIndices(element, depth + 1);
        if (depth) encodedIndices = TextNode.delimit(encodedIndices);
        result += encodedIndices;
      }
      const isNotLastElement = i < arr.length - 1;
      const separator = depth === 0 ? enclose(DELIMIT.separator, { opening: ' ', closing: ' ' }) : DELIMIT.separator;
      if (isNotLastElement && (!(Array.isArray(arr[i + 1])) || depth === 0)) result += separator;
    }

    return result;
  }

  static fromContent(content: string): TextNode | string {
    if (TokenNode.isDelimited(content)) return new VariantNode(content);
    else if (TokenNode.hasDelimited(content)) return new TokenNode(content);
    return content;
  }

  static hasDelimited(text: string): boolean {
    return text.includes(DELIMIT.opening) || text.includes(DELIMIT.closing);
  }

  static isDelimited(text: string): boolean {
    return text.startsWith(DELIMIT.opening) && text.endsWith(DELIMIT.closing);
  }

  assertAllIndicesConsumed(indices: number[], depth: number): void | never {
    if (depth === 0 && indices.length > 0) {
      throw new Error(`Unused variant indices. Received ${this.variantIndexCount}, leaving ${indices.length} unused.`);
    }
  }

  decodeIndices(encodedIndices: string | VariantIndices, depth: Integer = 0): Integer[] {
    const decodedIndices = TextNode.decodeIndices(encodedIndices);

    // Store the initial length of the indices array for later use in an error message
    if (depth === 0) this.variantIndexCount = decodedIndices.length;

    return decodedIndices;
  }

  abstract pick(options?: { seed?: Seed | undefined }): string;

  abstract pickIndices(
    options?: { depth?: Integer; indices?: VariantIndices | undefined; seed?: Seed | undefined },
  ): VariantIndices;

  abstract selectVariants(indices: string | VariantIndices, options?: { depth?: Integer }): string;

  abstract toString(): string;
}

class TokenNode extends TextNode {
  children?: (TextNode | string)[];

  constructor(content: string) {
    super(content);
    if (TextNode.hasDelimited(content)) {
      this.children = segmentByDelimited(content, DELIMIT).map((segment) => TextNode.fromContent(segment));
    }
  }

  pick(options: { seed?: Seed | undefined } = {}): string {
    const seed = IntSeededRng.spawn(options.seed);
    if (this.children) {
      return this.children.map((child) => typeof child === 'string' ? child : child.pick({ seed })).join('');
    }
    return this.content;
  }

  pickIndices(options: { depth?: Integer; seed?: Seed | undefined } = {}): VariantIndices {
    const { depth = 0 } = options;
    const seed = IntSeededRng.spawn(options.seed);

    const indices: VariantIndices = [];

    (this.children || [])
      .filter((child): child is VariantNode => child instanceof VariantNode)
      .forEach((variant) => {
        const childIndices = variant.pickIndices({ depth: depth + 1, seed });
        indices.push(
          childIndices.length === 1 ? childIndices[0] : childIndices,
        );
      });

    // Avoid unnecessary nesting
    return (indices.length === 1 && Array.isArray(indices[0])) ? indices[0] : indices;
  }

  selectVariants(indices: string | VariantIndices, options: { depth?: Integer } = {}): string {
    const { depth = 0 } = options;

    const decodedIndices = this.decodeIndices(indices, depth);
    const resolvedContent = (() => {
      if (!this.children) {
        return '';
      }

      const childContents = this.children.map((child) => {
        if (typeof child === 'string') return child;
        return child.selectVariants(decodedIndices, { depth: depth + 1 });
      });
      return childContents.join('');
    })();

    this.assertAllIndicesConsumed(decodedIndices, depth);
    return resolvedContent;
  }

  toString(): string {
    if (this.children) {
      return this.children.map((child) => child.toString()).join('');
    }
    return this.content;
  }
}

export class VariantNode extends TextNode {
  variants: (TextNode | string)[] = [];

  constructor(content: string) {
    super(content);
    this.variants = splitDelimited(content, DELIMIT).map((variant) => TextNode.fromContent(variant));
  }

  // Choose a variant at random
  pick(options: { seed?: Seed | undefined } = {}): string {
    const seed = IntSeededRng.spawn(options.seed);

    const index = pickInteger({ max: this.variants.length - 1, seed });

    const variant = this.variants[index];
    if (typeof variant === 'string') return variant;
    return variant.pick({ seed });
  }

  pickIndices(options: { depth?: Integer; seed?: Seed | undefined } = {}): VariantIndices {
    const { depth = 0 } = options;
    const seed = IntSeededRng.spawn(options.seed);

    const index = pickInteger({ max: this.variants.length - 1, seed });
    const nodeOrString = this.variants[index];

    const indices: VariantIndices = [index]; // index of the picked variant

    if (nodeOrString instanceof TextNode) {
      const childIndices = nodeOrString.pickIndices({ depth: depth + 1, seed });
      if (childIndices.length > 0) indices.push(childIndices);
    }

    return indices;
  }

  selectVariants(indices: VariantIndices, options: { depth?: Integer } = {}): string {
    const { depth = 0 } = options;
    const decodedIndices = this.decodeIndices(indices, depth);
    const variantIndex = decodedIndices.shift();

    if (variantIndex === undefined) throw new Error('Not enough indices to resolve all variants.');

    const selectedVariant = this.variants[variantIndex];
    if (selectedVariant === undefined) {
      throw new RangeError(
        `Variant index exceeds maximum index. Expected maximum of ${this.variants.length - 1}, got ${variantIndex}.`,
      );
    }

    const resolvedContent = (() => {
      if (typeof selectedVariant === 'string') return selectedVariant;
      return selectedVariant.selectVariants(decodedIndices, { depth: depth + 1 });
    })();

    this.assertAllIndicesConsumed(decodedIndices, depth);
    return resolvedContent;
  }

  toString(): string {
    return TextNode.delimit(this.variants.map((variant) => variant.toString()).join(DELIMIT.separator));
  }
}

function enclose(content: string, options: { opening: string; closing: string }): string {
  const { opening, closing } = options;
  return opening + content + closing;
}

function flatten(indices: VariantIndices): number[] {
  let flatIndices: number[] = [];
  for (const index of indices) {
    if (Array.isArray(index)) {
      flatIndices = flatIndices.concat(flatten(index));
    } else {
      flatIndices.push(index);
    }
  }
  return flatIndices;
}

function isNumberArray(items: unknown[]): items is number[] {
  return Array.isArray(items) && items.every((item) => typeof item === 'number');
}

type Integer = number;

type VariantIndices = (Integer | VariantIndices)[];
