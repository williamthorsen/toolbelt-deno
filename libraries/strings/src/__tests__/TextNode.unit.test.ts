import {
  assertEquals,
  assertMatch,
  assertObjectMatch,
  assertSnapshot,
  assertThrows,
  describe,
  it,
} from '../../dev_deps.ts';

import { TextNode, VariantNode } from '../TextNode.ts';

describe('TextNode', () => {
  // These are all equivalent inputs to `selectVariants`
  const nestedIndices = [[0, [1, [1]]], [1, [0, 0]]];
  const encodedIndices = '0[1[1]] | 1[0|0]';
  const flatIndices = [0, 1, 1, 1, 0, 0];

  describe('static create()', () => {
    it('if given a simple phrase, returns a single TextNode', () => {
      const input = 'Hello';
      const expected = {
        content: 'Hello',
        children: undefined,
      };

      const actual = TextNode.create(input);

      assertObjectMatch(actual, expected);
    });

    it('if given a delimited phrase, splits it into variants', () => {
      const input = '[A|B|C]';
      const expected = {
        content: input,
        variants: ['A', 'B', 'C'],
      };

      const actual = TextNode.create(input);

      assertObjectMatch(actual, expected);
    });

    it('given a string containing delimited phrases, splits it into tokens and variants', () => {
      const input = 'static [A|B] more static [C|D]';
      const expected = {
        content: input,
        children: [
          'static ',
          {
            content: '[A|B]',
            variants: ['A', 'B'],
          },
          ' more static ',
          {
            content: '[C|D]',
            variants: ['C', 'D'],
          },
        ],
      };

      const actual = TextNode.create(input);

      assertObjectMatch(actual, expected);
    });

    it('can handle nested delimiters', async (snapshot) => {
      const input = '1:[A[1[a|b]|2[c|d]]|B] 2:[C|D]';

      const actual = TextNode.create(input);

      await assertSnapshot(snapshot, actual);
    });
  });

  describe('static decodeIndices()', () => {
    it('given a string, splits on non-numeric sequences and returns an array of numbers', () => {
      const input = encodedIndices;
      const expected = flatIndices;

      const actual = TextNode.create('').decodeIndices(input);

      assertEquals(actual, expected);
    });

    it('flattens nested arrays', () => {
      const input = nestedIndices;
      const expected = flatIndices;

      const actual = TextNode.create('').decodeIndices(input);

      assertEquals(actual, expected);
    });

    describe('static encodeIndices()', () => {
      it('transforms an arbitrarily nested array of numbers into a string', () => {
        const input = nestedIndices;
        const expectedEncoded = encodedIndices;

        const actualEncoded = VariantNode.encodeIndices(input);

        assertEquals(actualEncoded, expectedEncoded);

        // Verify that the encoded string is equivalent to the input string
        assertEquals(TextNode.decodeIndices(actualEncoded), TextNode.decodeIndices(input));
      });
    });
  });

  describe('pick()', () => {
    it('reconstructs static content', () => {
      const input = 'Hello, world!';
      const expected = input;

      const actual = TextNode.create(input).pick();

      assertEquals(actual, expected);
    });

    it('randomly resolves variants', () => {
      const input = '[A|B|C]';
      const expected = /^[ABC]$/;

      const actual = TextNode.create(input).pick();

      assertMatch(actual, expected);
    });

    it('given the same seed, always returns the same result', () => {
      const seed = 1236;
      const input = 'token1 [A[1[a|b]|2[c|d]]|B] token2 [C|D[1|2]]';
      const knownExpected = 'token1 A2c token2 D2';

      const actual = TextNode.create(input).pick({ seed });

      assertEquals(actual, knownExpected);
    });
  });

  describe('pickIndices', () => {
    const input = '1:[A[1[a|b]|2[c|d]]|B] | 2:[C|D[1|2][a|b]|E]';
    const testCases = [
      { seed: 1234, expectedIndices: [1, 2], expectedString: '1:B | 2:E' },
      { seed: 1236, expectedIndices: [[0, [1, [0]]], [1, [1, 1]]], expectedString: '1:A2c | 2:D2b' },
      { seed: 1249, expectedIndices: [[0, [1, [1]]], [1, [0, 0]]], expectedString: '1:A2d | 2:D1a' },
    ];
    for (const { seed, expectedIndices, expectedString } of testCases) {
      it('selects depth-first indices to resolve variants', () => {
        const actualIndices = TextNode.create(input).pickIndices({ seed });
        const actualString = TextNode.create(input).pick({ seed });

        assertEquals(actualString, expectedString);
        assertEquals(actualIndices, expectedIndices);
      });
    }

    it('given a node without children, returns an empty array', () => {
      const input = 'Hello';
      const expected: number[] = [];

      const actual = TextNode.create(input).pickIndices();

      assertEquals(actual, expected);
    });
  });

  describe('selectVariants()', () => {
    it('consumes indices to select variants depth-first', () => {
      const input = '1:[A[1[a|b]|2[c|d]]|B] | 2:[C|D]';
      const textNode = TextNode.create(input);

      const indices = [0, 1, 1, 1];
      const nestedIndices = [[0, [1, [1]]], 1];
      const encoding = '0[1[1]]] | 1';
      const expected = '1:A2d | 2:D';

      const actualFromIndices = textNode.selectVariants(indices);
      const actualFromNestedIndices = textNode.selectVariants(nestedIndices);
      const actualFromEncoding = textNode.selectVariants(encoding);

      assertEquals(actualFromIndices, expected);
      assertEquals(actualFromNestedIndices, expected);
      assertEquals(actualFromEncoding, expected);
    });

    it('works when top-level node is a VariantNode', () => {
      const input = '[A|B[1|2]|C]';
      const indices = '1[0]';
      const ast = TextNode.create(input);
      const expected = 'B1';

      const actual = ast.selectVariants(indices);

      assertEquals(actual, expected);
    });

    it('given a text without children, returns an empty string', () => {
      const input = '';
      const expected = '';

      const actual = TextNode.create(input).selectVariants([]);

      assertEquals(actual, expected);
    });

    it('given an invalid index, throws an error', () => {
      const input = '[A|B|C]';
      const indices = [3];
      const ast = TextNode.create(input);

      const throwingFn = () => ast.selectVariants(indices);

      assertThrows(throwingFn, RangeError, 'Variant index exceeds maximum index. Expected maximum of 2, got 3.');
    });

    it('given too few indices to resolve all variants, throws an error', () => {
      const input = '[A[1|2]|B[3|4]|C[5|6]]'; // needs 2 indices to resolve
      const indices = [0];
      const ast = TextNode.create(input);

      const throwingFn = () => ast.selectVariants(indices);

      assertThrows(throwingFn, Error, 'Not enough indices to resolve all variants.');
    });

    it('given too many indices, throws an error', () => {
      const input = '[A|B|C]';
      const indices = [1, 1];
      const ast = TextNode.create(input);

      const throwingFn = () => ast.selectVariants(indices);

      assertThrows(throwingFn, Error, 'Unused variant indices. Received 2, leaving 1 unused.');
    });
  });

  describe('toString()', () => {
    it('reconstructs the original content', () => {
      const input = 'token1 [A[1[a|b]|2[c|d]]|B] token2 [C|D]';
      const expected = input;

      const actual = TextNode.create(input).toString();

      assertEquals(actual, expected);
    });

    it('given a text without children, returns an empty string', () => {
      const input = '';
      const expected = '';

      const actual = TextNode.create(input).toString();

      assertEquals(actual, expected);
    });
  });
});

describe('VariantNode class', () => {
});
