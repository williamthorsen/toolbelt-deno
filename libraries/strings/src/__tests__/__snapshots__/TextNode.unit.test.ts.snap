export const snapshot = {};

snapshot[`TextNode > static create() > can handle nested delimiters 1`] = `
TokenNode {
  children: [
    "1:",
    VariantNode {
      content: "[A[1[a|b]|2[c|d]]|B]",
      variantIndexCount: 0,
      variants: [
        TokenNode {
          children: [
            "A",
            VariantNode {
              content: "[1[a|b]|2[c|d]]",
              variantIndexCount: 0,
              variants: [
                TokenNode {
                  children: [
                    "1",
                    VariantNode {
                      content: "[a|b]",
                      variantIndexCount: 0,
                      variants: [
                        "a",
                        "b",
                      ],
                    },
                  ],
                  content: "1[a|b]",
                  variantIndexCount: 0,
                },
                TokenNode {
                  children: [
                    "2",
                    VariantNode {
                      content: "[c|d]",
                      variantIndexCount: 0,
                      variants: [
                        "c",
                        "d",
                      ],
                    },
                  ],
                  content: "2[c|d]",
                  variantIndexCount: 0,
                },
              ],
            },
          ],
          content: "A[1[a|b]|2[c|d]]",
          variantIndexCount: 0,
        },
        "B",
      ],
    },
    " 2:",
    VariantNode {
      content: "[C|D]",
      variantIndexCount: 0,
      variants: [
        "C",
        "D",
      ],
    },
  ],
  content: "1:[A[1[a|b]|2[c|d]]|B] 2:[C|D]",
  variantIndexCount: 0,
}
`;
